manticore/manticore.py

```diff
@@ -505,7 +537,193 @@ def load_assertions(self, path):
                     logger.debug("Repeated PC in assertions file %s", path)
                 self._assertions[pc] = ' '.join(line.split(' ')[1:])
 
+    def _store_state_callback(self, state, state_id):
+        logger.debug("store state %r", state_id)
+
+    def _load_state_callback(self, state, state_id):
+        logger.debug("load state %r", state_id)
+
+    def _terminate_state_callback(self, state, state_id, ex):
+        logger.info("Terminate_state_callback for %d", state_id)
+        executor = self._executor
+        #aggregates state statistics into exceutor statistics. FIXME split
+        logger.debug("Terminate state %r %r ", state, state_id)
+        state_visited = state.context.get('visited_since_last_fork', set())
+        state_instructions_count = state.context.get('instructions_count', 0)
+        with self._executor.locked_context() as executor_context:
+            executor_visited = executor_context.get('visited', set())
+            executor_context['visited'] = executor_visited.union(state_visited)
+
+            executor_instructions_count = executor_context.get('instructions_count', 0)
+            executor_context['instructions_count'] = executor_instructions_count + state_instructions_count 
+
+    def _fork_state_callback(self, state, expression, values, policy):
+        state_visited = state.context.get('visited_since_last_fork', set())
+        with self._executor.locked_context() as executor_context:
+            executor_visited = executor_context.get('visited', set())
+            executor_context['visited'] = executor_visited.union(state_visited)
+        state.context['visited_since_last_fork'] = set()
+
+        logger.debug("About to store state %r %r %r", state, expression, values, policy)
+
+    def _read_register_callback(self, state, platform, cpu, reg_name, value):
+        logger.debug("Read Register %r %r", reg_name, value)
+
+    def _write_register_callback(self, state, platform, cpu, reg_name, value):
+        logger.debug("Write Register %r %r", reg_name, value)
+
+    def _read_memory_callback(self, state, platform, cpu, address, value, size):
+        logger.debug("Read Memory %r %r %r", address, value, size)
+
+    def _write_memory_callback(self, state, platform, cpu, address, value, size):
+        logger.debug("Write Memory %r %r %r", address, value, size)
 
+    def _decode_instruction_callback(self, state, platform, cpu):
+        logger.debug("Decoding stuff instruction not available")
+
+
+    def _emulate_instruction_callback(self, state, platform, cpu, instruction):
+        logger.debug("About to emulate instruction")
+
+    def _did_execute_instruction_callback(self, state, platform, cpu, instruction):
+        logger.debug("Did execute an instruction")
+
+    def _execute_instruction_callback(self, state, platform, cpu, instruction):
```

Comment:
Forwarding Objects are saved internally at Signal anyway. We could omit them in them in the emit()
We can add an extra keyword argument to Signal.when() to control if the forwarding object has to interpose itself in the callback arguments or not.  
Idea is to have only state + specific arguments at the callback, right?

---- 
For the global event system. You mean that every object holds a reference to .events of type Events()
class Events():
   register_new_signal(str, desc)
   register_callback(str, obj, method)
   list_signals()
   disable()
   enable()

?